File: app/template_resolver.py
import requests as rq


class TemplateResolver:
    def __init__(self, app):
        self.app = app
        self.cache = {}

    def resolve(self, url):
        logger = self.app.logger
        if url not in self.cache:
            logger.info(f"Fetching template from: {url}")
            try:
                res = rq.get(url)
                if res.status_code == rq.codes.ok:
                    self.cache[url] = res.text
                else:
                    raise Exception(f"Received non OK code: {res.status_code}")
            except Exception as e:
                logger.error(f"Failed to fetch a template: {e}")
                raise
        return self.cache[url]

File: smtpie-api/src/main/java/smtpie/service/TemplateResolver.java
package smtpie.service;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.IOUtils;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

@Slf4j
public class TemplateResolver {

    private Map<String, String> templateCache = new HashMap<>();

    private HttpClient httpClient = HttpClients.createDefault();

    public String resolve(String url) {
        if (!templateCache.containsKey(url)) {
            templateCache.put(url, fetch(url));
        }
        return templateCache.get(url);
    }

    private String fetch(String url) {
        try {
            log.info("Fetching template from {}", url);
            HttpResponse res = httpClient.execute(new HttpGet(url));
            if (res.getStatusLine().getStatusCode() == HttpURLConnection.HTTP_OK) {
                return EntityUtils.toString(res.getEntity());
            } else {
                throw new RuntimeException(res.getStatusLine().toString());
            }
        } catch (Exception e) {
            log.error("Failed to fetch template: " + e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

}
